## Table of Contents
<ul>
<li>01. Importing libraries</li>
<li>02. Importing data</li>
<li>03. Harmonizing and merging data</li>
<li>04. Exporting</li>
</ul>

## 01. Importing Libraries

In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing data

In [6]:
path  = r'C:\Users\timol\Dropbox\CareerFoundry\Exercises\Achievement 6\WHR Analysis'

In [7]:
#Importing all three datasets
whr_df = pd.read_csv(os.path.join(path, '01 Data', 'Prepared Data', 'WHR-2024-12-02.csv'), index_col=False)
eiu_df = pd.read_csv(os.path.join(path, '01 Data', 'Prepared Data', 'EIU-2024-12-02.csv'), index_col=False)
wdi_df = pd.read_csv(os.path.join(path, '01 Data', 'Prepared Data', 'WDI-2024-12-02.csv'), index_col=False)

In [8]:
whr_df

Unnamed: 0      Country  Year  Happiness  GDP_log  Social_support  \
0              0  Afghanistan  2008      3.724    7.350           0.451   
1              1  Afghanistan  2009      4.402    7.509           0.552   
2              2  Afghanistan  2010      4.758    7.614           0.539   
3              3  Afghanistan  2011      3.832    7.581           0.521   
4              4  Afghanistan  2012      3.783    7.661           0.521   
...          ...          ...   ...        ...      ...             ...   
2358        2358     Zimbabwe  2019      2.694    7.698           0.759   
2359        2359     Zimbabwe  2020      3.160    7.596           0.717   
2360        2360     Zimbabwe  2021      3.155    7.657           0.685   
2361        2361     Zimbabwe  2022      3.296    7.670           0.666   
2362        2362     Zimbabwe  2023      3.572    7.679           0.694   

      Life_expectancy  Freedom  Generosity  Corruption  Pos_affect  Neg_affect  
0              50.500    0.718       0.164       0.882       0.414       0.258  
1              50.800    0.679       0.187       0.850       0.481       0.237  
2              51.100    0.600       0.118       0.707       0.517       0.275  
3              51.400    0.496       0.160       0.731       0.480       0.267  
4              51.700    0.531       0.234       0.776       0.614       0.268  
...               ...      ...         ...         ...         ...         ...  
2358           53.100    0.632      -0.051       0.831       0.658       0.235  
2359           53.575    0.643       0.003       0.789       0.661       0.346  
2360           54.050    0.668      -0.079       0.757       0.610       0.242  
2361           54.525    0.652      -0.073       0.753       0.641       0.191  
2362           55.000    0.735      -0.069       0.757       0.610       0.179  

[2363 rows x 12 columns]

In [9]:
eiu_df

Unnamed: 0      Country  Year  Democracy
0              0  Afghanistan  2006       3.06
1              1  Afghanistan  2008       3.02
2              2  Afghanistan  2010       2.48
3              3  Afghanistan  2011       2.48
4              4  Afghanistan  2012       2.48
...          ...          ...   ...        ...
2500        2500     Zimbabwe  2018       3.16
2501        2501     Zimbabwe  2019       3.16
2502        2502     Zimbabwe  2020       3.16
2503        2503     Zimbabwe  2021       2.92
2504        2504     Zimbabwe  2022       2.92

[2505 rows x 4 columns]

In [10]:
wdi_df

Unnamed: 0      Country  Year   Inflation
0              0  Afghanistan  2008   26.418664
1              1  Afghanistan  2009   -6.811161
2              2  Afghanistan  2010    2.178538
3              3  Afghanistan  2011   11.804186
4              4  Afghanistan  2012    6.441213
...          ...          ...   ...         ...
3586        3586     Zimbabwe  2018   10.618866
3587        3587     Zimbabwe  2019  255.304991
3588        3588     Zimbabwe  2020  557.201817
3589        3589     Zimbabwe  2021   98.546105
3590        3590     Zimbabwe  2022  104.705171

[3591 rows x 4 columns]

In [11]:
#Removing index column in all three CSVs
whr_df = whr_df.iloc[:, 1:]
eiu_df = eiu_df.iloc[:, 1:]
wdi_df = wdi_df.iloc[:, 1:]

In [12]:
whr_df

Country  Year  Happiness  GDP_log  Social_support  Life_expectancy  \
0     Afghanistan  2008      3.724    7.350           0.451           50.500   
1     Afghanistan  2009      4.402    7.509           0.552           50.800   
2     Afghanistan  2010      4.758    7.614           0.539           51.100   
3     Afghanistan  2011      3.832    7.581           0.521           51.400   
4     Afghanistan  2012      3.783    7.661           0.521           51.700   
...           ...   ...        ...      ...             ...              ...   
2358     Zimbabwe  2019      2.694    7.698           0.759           53.100   
2359     Zimbabwe  2020      3.160    7.596           0.717           53.575   
2360     Zimbabwe  2021      3.155    7.657           0.685           54.050   
2361     Zimbabwe  2022      3.296    7.670           0.666           54.525   
2362     Zimbabwe  2023      3.572    7.679           0.694           55.000   

      Freedom  Generosity  Corruption  Pos_affect  Neg_affect  
0       0.718       0.164       0.882       0.414       0.258  
1       0.679       0.187       0.850       0.481       0.237  
2       0.600       0.118       0.707       0.517       0.275  
3       0.496       0.160       0.731       0.480       0.267  
4       0.531       0.234       0.776       0.614       0.268  
...       ...         ...         ...         ...         ...  
2358    0.632      -0.051       0.831       0.658       0.235  
2359    0.643       0.003       0.789       0.661       0.346  
2360    0.668      -0.079       0.757       0.610       0.242  
2361    0.652      -0.073       0.753       0.641       0.191  
2362    0.735      -0.069       0.757       0.610       0.179  

[2363 rows x 11 columns]

In [13]:
eiu_df

Country  Year  Democracy
0     Afghanistan  2006       3.06
1     Afghanistan  2008       3.02
2     Afghanistan  2010       2.48
3     Afghanistan  2011       2.48
4     Afghanistan  2012       2.48
...           ...   ...        ...
2500     Zimbabwe  2018       3.16
2501     Zimbabwe  2019       3.16
2502     Zimbabwe  2020       3.16
2503     Zimbabwe  2021       2.92
2504     Zimbabwe  2022       2.92

[2505 rows x 3 columns]

In [14]:
wdi_df

Country  Year   Inflation
0     Afghanistan  2008   26.418664
1     Afghanistan  2009   -6.811161
2     Afghanistan  2010    2.178538
3     Afghanistan  2011   11.804186
4     Afghanistan  2012    6.441213
...           ...   ...         ...
3586     Zimbabwe  2018   10.618866
3587     Zimbabwe  2019  255.304991
3588     Zimbabwe  2020  557.201817
3589     Zimbabwe  2021   98.546105
3590     Zimbabwe  2022  104.705171

[3591 rows x 3 columns]

## 03. Harmonizing and merging data

### 03.01 Harmonizing the 'country' variable

The WDI data contains a lot of geographic entities that are not actually countries, such as 'East Asia & Pacific' or 'Early-demographic dividend'. These should be removed from the dataset before mergin with the other dataframes. To identify the non-country entities I loaded the CSV into Claude and asked it to produce a list of non-country entities. I then adjusted this list manually to keep entities like 'West Bank and Gaza' or 'Taiwan'. 

In [18]:
# List of non-country designations to remove
non_countries = [
    # Regional groupings
    'East Asia & Pacific',
    'East Asia & Pacific (excluding high income)',
    'East Asia & Pacific (IDA & IBRD countries)',
    'Europe & Central Asia',
    'Europe & Central Asia (excluding high income)',
    'Europe & Central Asia (IDA & IBRD countries)',
    'Euro area',
    'European Union',
    'Latin America & Caribbean',
    'Latin America & Caribbean (excluding high income)',
    'Latin America & the Caribbean (IDA & IBRD countries)',
    'Middle East & North Africa',
    'Middle East & North Africa (excluding high income)',
    'Middle East & North Africa (IDA & IBRD countries)',
    'North America',
    'South Asia',
    'South Asia (IDA & IBRD)',
    'Sub-Saharan Africa',
    'Sub-Saharan Africa (excluding high income)',
    'Sub-Saharan Africa (IDA & IBRD countries)',
    'Central Europe and the Baltics',
    'World',
    
    # Development status groups
    'Early-demographic dividend',
    'Post-demographic dividend',
    'Late-demographic dividend',
    'High income',
    'Low & middle income',
    'Low income',
    'Lower middle income',
    'Upper middle income',
    'Heavily indebted poor countries (HIPC)',
    'Least developed countries: UN classification',
    'Small states',
    'Other small states',
    'Pacific island small states',
    
    # Other
    'Macao SAR, China',
    
    # IDA/IBRD groupings
    'IBRD only',
    'IDA & IBRD total',
    'IDA blend',
    'IDA only',
    'IDA total',
    
    # Ensure any spaces or variations are caught
    'IDA & IBRD',
    'IBRD',
    'IDA'
]

In [19]:
non_countries

['East Asia & Pacific',
 'East Asia & Pacific (excluding high income)',
 'East Asia & Pacific (IDA & IBRD countries)',
 'Europe & Central Asia',
 'Europe & Central Asia (excluding high income)',
 'Europe & Central Asia (IDA & IBRD countries)',
 'Euro area',
 'European Union',
 'Latin America & Caribbean',
 'Latin America & Caribbean (excluding high income)',
 'Latin America & the Caribbean (IDA & IBRD countries)',
 'Middle East & North Africa',
 'Middle East & North Africa (excluding high income)',
 'Middle East & North Africa (IDA & IBRD countries)',
 'North America',
 'South Asia',
 'South Asia (IDA & IBRD)',
 'Sub-Saharan Africa',
 'Sub-Saharan Africa (excluding high income)',
 'Sub-Saharan Africa (IDA & IBRD countries)',
 'Central Europe and the Baltics',
 'World',
 'Early-demographic dividend',
 'Post-demographic dividend',
 'Late-demographic dividend',
 'High income',
 'Low & middle income',
 'Low income',
 'Lower middle income',
 'Upper middle income',
 'Heavily indebted poor co

In [20]:
#Create 'country_mask' that does NOT contain list 'non_countries'
country_mask = ~wdi_df['Country'].isin(non_countries)

In [21]:
country_mask

0       True
1       True
2       True
3       True
4       True
        ... 
3586    True
3587    True
3588    True
3589    True
3590    True
Name: Country, Length: 3591, dtype: bool

In [22]:
# Apply the mask to get only country entries
wdi_cleaned_df = wdi_df[country_mask].copy()

In [23]:
wdi_cleaned_df

Country  Year   Inflation
0     Afghanistan  2008   26.418664
1     Afghanistan  2009   -6.811161
2     Afghanistan  2010    2.178538
3     Afghanistan  2011   11.804186
4     Afghanistan  2012    6.441213
...           ...   ...         ...
3586     Zimbabwe  2018   10.618866
3587     Zimbabwe  2019  255.304991
3588     Zimbabwe  2020  557.201817
3589     Zimbabwe  2021   98.546105
3590     Zimbabwe  2022  104.705171

[2937 rows x 3 columns]

We reduced the df from 3591 rows to 2937 rows.

Check for additional variations between country names in both dataframes

In [26]:
# Get unique country names from each dataframe
eiu_countries = set(eiu_df['Country'].unique())
wdi_countries = set(wdi_cleaned_df['Country'].unique())

In [27]:
# Find countries that are in EIU but not in cleaned WDI
only_in_eiu = eiu_countries - wdi_countries

In [28]:
sorted(only_in_eiu)  # sorted() makes the output easier to read

['Argentina',
 'Cuba',
 'Eritrea',
 "Korea, Dem. People's Rep.",
 'Taiwan, China',
 'Turkmenistan']

In [29]:
# Find countries that are in WDI but not in EIU
only_in_wdi = wdi_countries - eiu_countries
sorted(only_in_wdi)

['Antigua and Barbuda',
 'Aruba',
 'Bahamas, The',
 'Barbados',
 'Belize',
 'Brunei Darussalam',
 'Cayman Islands',
 'Curacao',
 'Dominica',
 'Fragile and conflict affected situations',
 'Grenada',
 'Kiribati',
 'Kosovo',
 'Maldives',
 'Micronesia, Fed. Sts.',
 'Middle income',
 'Nauru',
 'New Caledonia',
 'OECD members',
 'Palau',
 'Pre-demographic dividend',
 'Samoa',
 'San Marino',
 'Sao Tome and Principe',
 'Seychelles',
 'Sint Maarten (Dutch part)',
 'Solomon Islands',
 'South Sudan',
 'St. Kitts and Nevis',
 'St. Lucia',
 'St. Vincent and the Grenadines',
 'Tonga',
 'Tuvalu',
 'Vanuatu']

WDI still contains more entities that EIU. But we can see that there are no variations of spellings, which would be the bigger issue. We also know from looking at the source data in Excel that the countrys that are only part of EIU area actually missing from the WDI dataset. A left join should merge all neccessary data at this point.

### 03.02. Merging EIU and WDI data

In [32]:
#Checking dtypes
eiu_df.dtypes

Country       object
Year           int64
Democracy    float64
dtype: object

In [33]:
wdi_cleaned_df.dtypes

Country       object
Year           int64
Inflation    float64
dtype: object

In [34]:
# Perform left join on Country and Year for EIU and WDI
merged_eiu_wdi_df = pd.merge(
    left=eiu_df,
    right=wdi_cleaned_df,
    how='left',
    on=['Country', 'Year']
)

In [35]:
merged_eiu_wdi_df

Country  Year  Democracy   Inflation
0     Afghanistan  2006       3.06         NaN
1     Afghanistan  2008       3.02   26.418664
2     Afghanistan  2010       2.48    2.178538
3     Afghanistan  2011       2.48   11.804186
4     Afghanistan  2012       2.48    6.441213
...           ...   ...        ...         ...
2500     Zimbabwe  2018       3.16   10.618866
2501     Zimbabwe  2019       3.16  255.304991
2502     Zimbabwe  2020       3.16  557.201817
2503     Zimbabwe  2021       2.92   98.546105
2504     Zimbabwe  2022       2.92  104.705171

[2505 rows x 4 columns]

In [36]:
merged_eiu_wdi_df.describe()

Year   Democracy    Inflation
count  2505.000000  2505.00000  2175.000000
mean   2014.800000     5.46883     6.546811
std       4.651377     2.22480    18.984120
min    2006.000000     0.32000    -4.294873
25%    2011.000000     3.46000     1.628215
50%    2015.000000     5.77000     3.589220
75%    2019.000000     7.24000     7.041087
max    2022.000000     9.93000   557.201817

### 03.02. Merging WHR and EIU/WDI data

Checking for difference in country names between WHR and merged_eiu_wdi_df

In [39]:
# Get unique country names from each dataframe
merged_eiu_wdi_countries = set(merged_eiu_wdi_df['Country'].unique())
whr_countries = set(whr_df['Country'].unique())

In [40]:
# Find countries that are in WHR but not in merged_eiu_wdi
only_in_whr = whr_countries - merged_eiu_wdi_countries

In [41]:
sorted(only_in_whr)

['Belize',
 'Congo (Brazzaville)',
 'Congo (Kinshasa)',
 'Egypt',
 'Gambia',
 'Hong Kong S.A.R. of China',
 'Iran',
 'Ivory Coast',
 'Kosovo',
 'Kyrgyzstan',
 'Laos',
 'Maldives',
 'Russia',
 'Slovakia',
 'Somalia',
 'Somaliland region',
 'South Korea',
 'South Sudan',
 'State of Palestine',
 'Syria',
 'Taiwan Province of China',
 'Türkiye',
 'Venezuela',
 'Yemen']

In [42]:
# Find countries that are in merged_eiu_wdi but not in WHR
only_in_merge = merged_eiu_wdi_countries - whr_countries

In [43]:
sorted(only_in_merge)

['Cabo Verde',
 'Congo, Dem. Rep.',
 'Congo, Rep.',
 "Cote d'Ivoire",
 'Egypt, Arab Rep.',
 'Equatorial Guinea',
 'Eritrea',
 'Fiji',
 'Gambia, The',
 'Guinea-Bissau',
 'Hong Kong SAR, China',
 'Iran, Islamic Rep.',
 "Korea, Dem. People's Rep.",
 'Korea, Rep.',
 'Kyrgyz Republic',
 'Lao PDR',
 'Papua New Guinea',
 'Russian Federation',
 'Slovak Republic',
 'Syrian Arab Republic',
 'Taiwan, China',
 'Timor-Leste',
 'Turkiye',
 'Venezuela, RB',
 'West Bank and Gaza',
 'Yemen, Rep.']

In [44]:
#Replacing selected country names in whr_df
whr_df_1 = whr_df.replace(['State of Palestine', 'Taiwan Province of China', 'Congo (Brazzaville)', 'Congo (Kinshasa)', 'Hong Kong S.A.R. of China', 'Ivory Coast'],
['West Bank and Gaza', 'Taiwan, China', 'Congo, Rep', 'Congo, Dem. Rep.', 'Hong Kong SAR, China', 'Cote d\'Ivoire'])

In [45]:
#Replacing selected country names in merged_eiu_wdi_df
merged_eiu_wdi_df_1 = merged_eiu_wdi_df.replace(['Egypt, Arab Rep.', 'Gambia, The', 'Iran, Islamic Rep.', 'Korea, Rep.', 'Kyrgyz Republic', 'Lao PDR', 'Russian Federation', 'Slovak Republic', 'Syrian Arab Republic', 'Turkiye', 'Venezuela, RB', 'Yemen, Rep.'],
['Egypt', 'Gambia', 'Iran', 'South Korea',  'Kyrgyzstan', 'Laos', 'Russia',  'Slovakia',  'Syria',  'Türkiye', 'Venezuela', 'Yemen'])

In [46]:
# Perform left join on Country and Year for WHR and merged_eiu_wdi_df
merged_all_df = pd.merge(
    left=whr_df_1,
    right=merged_eiu_wdi_df_1,
    how='left',
    on=['Country', 'Year'], 
    indicator = True
)

In [47]:
merged_all_df

Country  Year  Happiness  GDP_log  Social_support  Life_expectancy  \
0     Afghanistan  2008      3.724    7.350           0.451           50.500   
1     Afghanistan  2009      4.402    7.509           0.552           50.800   
2     Afghanistan  2010      4.758    7.614           0.539           51.100   
3     Afghanistan  2011      3.832    7.581           0.521           51.400   
4     Afghanistan  2012      3.783    7.661           0.521           51.700   
...           ...   ...        ...      ...             ...              ...   
2358     Zimbabwe  2019      2.694    7.698           0.759           53.100   
2359     Zimbabwe  2020      3.160    7.596           0.717           53.575   
2360     Zimbabwe  2021      3.155    7.657           0.685           54.050   
2361     Zimbabwe  2022      3.296    7.670           0.666           54.525   
2362     Zimbabwe  2023      3.572    7.679           0.694           55.000   

      Freedom  Generosity  Corruption  Pos_affect  Neg_affect  Democracy  \
0       0.718       0.164       0.882       0.414       0.258       3.02   
1       0.679       0.187       0.850       0.481       0.237        NaN   
2       0.600       0.118       0.707       0.517       0.275       2.48   
3       0.496       0.160       0.731       0.480       0.267       2.48   
4       0.531       0.234       0.776       0.614       0.268       2.48   
...       ...         ...         ...         ...         ...        ...   
2358    0.632      -0.051       0.831       0.658       0.235       3.16   
2359    0.643       0.003       0.789       0.661       0.346       3.16   
2360    0.668      -0.079       0.757       0.610       0.242       2.92   
2361    0.652      -0.073       0.753       0.641       0.191       2.92   
2362    0.735      -0.069       0.757       0.610       0.179        NaN   

       Inflation     _merge  
0      26.418664       both  
1            NaN  left_only  
2       2.178538       both  
3      11.804186       both  
4       6.441213       both  
...          ...        ...  
2358  255.304991       both  
2359  557.201817       both  
2360   98.546105       both  
2361  104.705171       both  
2362         NaN  left_only  

[2363 rows x 14 columns]

In [48]:
merged_all_df.describe()

Year    Happiness      GDP_log  Social_support  Life_expectancy  \
count  2363.000000  2363.000000  2335.000000     2350.000000      2300.000000   
mean   2014.763860     5.483566     9.399671        0.809369        63.401828   
std       5.059436     1.125522     1.152069        0.121212         6.842644   
min    2005.000000     1.281000     5.527000        0.228000         6.720000   
25%    2011.000000     4.647000     8.506500        0.744000        59.195000   
50%    2015.000000     5.449000     9.503000        0.834500        65.100000   
75%    2019.000000     6.323500    10.392500        0.904000        68.552500   
max    2023.000000     8.019000    11.676000        0.987000        74.600000   

           Freedom   Generosity   Corruption   Pos_affect   Neg_affect  \
count  2327.000000  2282.000000  2238.000000  2339.000000  2347.000000   
mean      0.750282     0.000098     0.743971     0.651882     0.273151   
std       0.139357     0.161388     0.184865     0.106240     0.087131   
min       0.228000    -0.340000     0.035000     0.179000     0.083000   
25%       0.661000    -0.112000     0.687000     0.572000     0.209000   
50%       0.771000    -0.022000     0.798500     0.663000     0.262000   
75%       0.862000     0.093750     0.867750     0.737000     0.326000   
max       0.985000     0.700000     0.983000     0.884000     0.705000   

         Democracy    Inflation  
count  1943.000000  1768.000000  
mean      5.766454     6.203946  
std       2.119415    18.227375  
min       0.320000    -3.749145  
25%       3.945000     1.589898  
50%       5.990000     3.466150  
75%       7.520000     6.913544  
max       9.930000   557.201817

In [49]:
# Count NaNs in the entire DataFrame
nan_count_all = merged_all_df.isna().sum()
print("Number of NaNs in each column:")
nan_count_all

Number of NaNs in each column:


Country              0
Year                 0
Happiness            0
GDP_log             28
Social_support      13
Life_expectancy     63
Freedom             36
Generosity          81
Corruption         125
Pos_affect          24
Neg_affect          16
Democracy          420
Inflation          595
_merge               0
dtype: int64

## 04. Exporting data

In [51]:
#Export Data as Pickle
merged_all_df.to_pickle(os.path.join(path, '01 Data','Prepared Data', 'dataset-2024-02-12.pkl'))